internal/controller/pipelinerollout_controller.go

```diff
@@ -408,49 +433,105 @@ func mergePipeline(existingPipeline *kubernetes.GenericObject, newPipeline *kube
 	resultPipeline.Labels = maps.Clone(newPipeline.Labels)
 	return resultPipeline
 }
+
 func (r *PipelineRolloutReconciler) processExistingPipeline(ctx context.Context, pipelineRollout *apiv1.PipelineRollout,
 	existingPipelineDef *kubernetes.GenericObject, newPipelineDef *kubernetes.GenericObject, syncStartTime time.Time) error {
 
-	// Get the fields we need from both the Pipeline spec we have and the one we want
-	// TODO: consider having one struct which include our GenericObject plus our PipelineSpec so we can avoid multiple repeat conversions
+	numaLogger := logger.FromContext(ctx)
 
-	var existingPipelineSpec PipelineSpec
-	if err := json.Unmarshal(existingPipelineDef.Spec.Raw, &existingPipelineSpec); err != nil {
-		return fmt.Errorf("failed to convert existing Pipeline spec %q into PipelineSpec type, err=%v", string(existingPipelineDef.Spec.Raw), err)
+	// what is the preferred strategy for this namespace?
+	userPreferredStrategy, err := usde.GetUserStrategy(ctx, newPipelineDef.Namespace)
+	if err != nil {
+		return err
 	}
 
-	// Does pipeline spec need to be updated?
-	pipelineNeedsToUpdate, err := pipelineNeedsUpdating(ctx, newPipelineDef, existingPipelineDef)
+	// does the Resource need updating, and if so how?
+	pipelineNeedsToUpdate, upgradeStrategyType, err := usde.ResourceNeedsUpdating(ctx, newPipelineDef, existingPipelineDef)
 	if err != nil {
 		return err
 	}
+	numaLogger.
+		WithValues("pipelineNeedsToUpdate", pipelineNeedsToUpdate, "upgradeStrategyType", upgradeStrategyType).
+		Debug("Upgrade decision result")
 
-	// set the Status appropriately to "Pending" or "Deployed"
-	// if pipelineNeedsToUpdate - this means there's a mismatch between the desired Pipeline spec and actual Pipeline spec
+	// set the Status appropriately to "Pending" or "Deployed" depending on whether pipeline needs to update
 	if pipelineNeedsToUpdate {
 		pipelineRollout.Status.MarkPending()
 	} else {
 		pipelineRollout.Status.MarkDeployed(pipelineRollout.Generation)
 	}
 
-	if common.DataLossPrevention { // feature flag
-		if err = r.processExistingPipelineWithoutDataLoss(ctx, pipelineRollout, existingPipelineDef, newPipelineDef, pipelineNeedsToUpdate); err != nil {
-			return err
+	// is there currently an inProgressStrategy for the pipeline? (This will override any new decision)
+	inProgressStrategy := r.inProgressStrategyMgr.getStrategy(ctx, pipelineRollout)
+
+	// if not, should we set one?
+	// inProgressStrategy is used for PPND and Progressive strategies (i.e. any strategies which require multiple reconciliations to perform)
+	if userPreferredStrategy == config.PPNDStrategyID && inProgressStrategy != apiv1.UpgradeStrategyPPND {
+		// if the preferred strategy is PPND, do we need to start the process for PPND (if we haven't already)?
+		needPPND := false
+		if userPreferredStrategy == config.PPNDStrategyID {
+			ppndRequired, err := r.needPPND(ctx, pipelineRollout, newPipelineDef, upgradeStrategyType == apiv1.UpgradeStrategyPPND)
```

Comment:
We should not alter the runtime in progress upgrade strategy right? As that implies there is an upgrade with a certain strategy already going on.

Reply:
I guess you are referring to the case in which a user changes their preferred strategy on their namespace while something is already happening. You're right - I didn't take care of that case. 

Perhaps I can just change like 469 above to:
`if userPreferredStrategy == config.PPNDStrategyID && inProgressStrategy == apiv1.UpgradeStrategyNoOp`